egs/wsj/s5/utils/lang/make_subword_lexicon_fst.py

```diff
@@ -0,0 +1,159 @@
+
+    def is_end(word, separator):
+        return (not word.endswith(separator))
+
+    def get_suffix(phone):
+        return phone[-2:]
+
+    def get_position(phones):
+        phones = phones[:-1] if contain_disambig_symbol(phones) else phones
+
+        if len(phones) == 1:
+            suffix = get_suffix(phones[0])
+            if suffix == "_B":
+                return "begin"
+            elif suffix == "_I":
+                return "middle"
+            elif suffix == "_E":
+                return "end"
+            else:
+                return "word"
+        else:
+            if get_suffix(phones[0]) == "_B":
+                if get_suffix(phones[-1]) == "_I":
+                    return "begin"
+                elif get_suffix(phones[-1]) == "_E":
+                    return "word"
+            elif get_suffix(phones[0]) == "_I":
+                if get_suffix(phones[-1]) == "_I":
+                    return "middle"
+                elif get_suffix(phones[-1]) == "_E":
+                    return "end"
+
+    def print_word(word, pron_cost, phones, position, next_state):
+    
+        assert(len(phones)) >= 0
+    
+        from_state = prefix_state
+        to_state = prefix_state
+        if position == "begin":
+            from_state = loop_state
+        elif position == "end": 
+            to_state = "sil_nonsil"
+        elif position == "word":
+            from_state = loop_state
+            to_state = "sil_nonsil"
+    
+        current_state = from_state
+        for i in range(len(phones) - 1):
+            word = (word if i == 0 else '<eps>')
+            cost = (pron_cost if i == 0 else 0.0)
+            print_arc(current_state, next_state, phones[i], word, cost)
+            current_state = next_state
+            next_state += 1
+    
+        i = len(phones) - 1
+        phone = (phones[i] if i >= 0 else "<eps>")
+        word = (word if i <= 0 else "<eps>")
+        cost = (pron_cost if i <= 0 else 0.0)
+        if to_state != "sil_nonsil":
+            print_arc(current_state, to_state, phone, word, cost)
+        else:
+            p_cost = cost
+            cost = no_sil_cost + p_cost
+            print_arc(current_state, loop_state, phone, word, cost)
+            cost = sil_cost + p_cost
+            print_arc(current_state, sil_state, phone, word, cost)
+    
+        return next_state
+
+    sil_cost = -math.log(sil_prob)
+    no_sil_cost = -math.log(1 - sil_prob)
+
+    start_state = 0
+    loop_state = 1
+    sil_state = 2
+    prefix_state = 3
+    next_state = 4
+
+    print_arc(start_state, loop_state, '<eps>', '<eps>', no_sil_cost)
+    print_arc(start_state, sil_state, '<eps>', '<eps>', sil_cost)
+    if sil_disambig is None:
+        print_arc(sil_state, loop_state, sil_phone, '<eps>', 0.0)
+    else:
+        disambig_state = next_state
+        next_state += 1
+        print_arc(sil_state, disambig_state, sil_phone, '<eps>', 0.0)
```

Comment:
having things like '<eps>' everywhere would make it hard for, say for certain system we want to use <EPS> for epsilons. Maybe declare a const variable ESP = "<eps>" and pass that throughout the script.

Similarly for the "end" "word" etc.